File: src/cpp/multithreading/multiple_worker_threads.cpp
#include <iostream>
#include <thread>
#include <vector>

void foo(int id) {
  std::cout << "Function foo() starts, Thread id: " << id << std::endl;
  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
  std::cout << "Function foo() finishes, Thread id: " << id << std::endl;
}

void bar(int id) {
  std::cout << "Function bar() starts, Thread id: " << id << std::endl;
  std::this_thread::sleep_for(std::chrono::milliseconds(100));
  std::cout << "Function bar() finishes, Thread id: " << id << std::endl;
}

auto main() -> int {

  std::cout << "Main thread starts" << std::endl;

  std::vector<std::thread> threads;
  threads.reserve(5);
  for (int i = 0; i < 5; i++) {
    threads.emplace_back(foo, i);
  }

  for (int i = 0; i < 5; i++) {
    threads.emplace_back(bar, i);
  }

  for (auto &thread : threads) {
    thread.join();
  }

  std::cout << "Main thread finishes" << std::endl;

  return 0;
}

File: src/python/multithreading/multiple_worker_threads.py
"""
Multiple Worker Threads Example

This script demonstrates creating and managing multiple threads that execute
different functions concurrently. It illustrates:
- Creating multiple threads with different target functions
- Passing arguments to thread target functions
- Thread interleaving and concurrent execution

Key Concepts:
- Threads can run different functions with varying execution times
- Thread execution order is non-deterministic (depends on OS scheduling)
- Using join() ensures all threads complete before the main thread continues

Observe how the output interleaves as threads with different sleep times
execute concurrently.
"""

import time
from threading import Thread


def foo(thread_id):
    """A worker function with a 1-second execution time."""
    print(f"Function foo() starts, Thread id: {thread_id}")
    time.sleep(1)
    print(f"Function foo() finishes, Thread id: {thread_id}")


def bar(thread_id):
    """A worker function with a 0.1-second execution time."""
    print(f"Function bar() starts, Thread id: {thread_id}")
    time.sleep(0.1)
    print(f"Function bar() finishes, Thread id: {thread_id}")


def main():
    print("Main thread starts")

    threads = []

    # Create 5 threads running foo()
    for i in range(5):
        threads.append(Thread(target=foo, args=(i,)))

    # Create 5 threads running bar()
    for i in range(5):
        threads.append(Thread(target=bar, args=(i,)))

    # Start all threads
    for thread in threads:
        thread.start()

    # Wait for all threads to complete
    for thread in threads:
        thread.join()

    print("Main thread finishes")


if __name__ == "__main__":
    main()
